### DISCUSSION TITLE
How to performance-test a contract? Available tools?

### DISCUSSION THREAD

**Discussion by dckc** [Agoric/agoric-sdk#9941] [Q&A]
It's required by the checklist:
 - https://github.com/Agoric/agoric-sdk/discussions/8280

> Perform performance and stress testing on the smart contract deployed on our testnet. Build test tools and scripts to simulate high loads, monitor resource usage, and identify bottlenecks. Document results, including performance metrics, failure points, and recommendations for improvement.

Any suggestions on how to do it? Are there any available tools?

cc @tgrecojs

**Comment by dckc** [comment]
## Benchmarking Contract operations at the SwingSet VM Level

One somewhat mature set of tools for benchmarking smart contract operations runs within a SwingSet VM, but _not_ using the cosmos-sdk consensus layer:

 - [The Benchmarkerator](https://github.com/Agoric/agoric-sdk/blob/master/packages/benchmark/doc/benchmarkerator.md)

There's a somewhat extended write-up:

 - [Running the benchmark for vault creation](https://github.com/Agoric/agoric-sdk/blob/master/packages/benchmark/doc/usingBenchmarks/usingBenchmarks.md)

**Comment by dckc** [comment]
## end-to-end testing with transactions and queries

The ultimate test is to run an actual blockchain, submit transactions, and run queries.

For example, we have a shell script that tests minting a KREAd character:

```sh
KREAD_ITEM_OFFER=$(mktemp -t kreadItem.XXX)
node ./generate-kread-item-request.mjs > $KREAD_ITEM_OFFER
agops perf satisfaction --from $GOV1ADDR --executeOffer $KREAD_ITEM_OFFER --keyring-backend=test

agd query vstorage data published.wallet.$GOV1ADDR.current  -o json >& gov1.out
name=`jq '.value | fromjson | .values[2] | fromjson | .body[1:] | fromjson | .purses[1].balance.value.payload[0][0].name ' gov1.out`
test_val $name \"ephemeral_Ace\" "found KREAd character"
```
 -- https://github.com/Agoric/agoric-3-proposals/blob/main/proposals/71%3Aupgrade-14/create-kread-item-test.sh

`agops perf satisfaction` is a test tool that submits a smart-wallet transaction and polls until the offer is satisfied.

Before KREAd went into production, we generated various levels of load using scripts like that and looked at various metrics. I think we ran them not just in an a3p context, but on an actual multi-validator-node test network in a kubernetes cluster, and we looked at the output using tools such as datadog. @toliaqat is there any more detail to share in this context?

**Comment by dckc** [comment]
## end-to-end testing APIs

There are a few other tools for submitting transactions and making queries in the [@agoric/synthetic-chain](https://github.com/Agoric/agoric-3-proposals/tree/main/packages/synthetic-chain) package.

See also:

 - https://github.com/Agoric/agoric-sdk/issues/8963

... including recent `multichain-testing` work such as...

```js
    const wdUser1 = await provisionSmartWallet(agoricAddr, {
      BLD: 100_000n,
      IST: 100_000n,
    });
    t.log(`provisioning agoric smart wallet for ${agoricAddr}`);

    const doOffer = makeDoOffer(wdUser1);

    const brands = await vstorageClient.queryData(
      'published.agoricNames.brand',
    );
```
 -- https://github.com/Agoric/agoric-sdk/blob/master/multichain-testing/test/send-anywhere.test.ts

**Comment by erights** [comment]
Attn @gibson042

**Comment by dckc** [comment]
Here's a goal from internal discussion. I don't expect that this is self-explanatory, but I'll go ahead and share it...

 - tool which submits txns at a fixed rate, and records completion time, and displays latency and acceptance rate, and some kind of "how busy is the chain" metric like block times or duration of the `cosmic-swingset-begin-block` -> `cosmic-swingset-end-block-finish` elapsed time
 - run that at maybe 1txn/min for 10-15min, then increase to 2/min, then 3/min, etc, and plot the curves

we should learn how many computrons are spent to do one of those offers, to guess what our 65Mc limit will do to the scheduling.

Once we deploy to a multi-node net, we should see how it compares against the actual wallclock time needed by those validators.

----

p.s.

65Mc = 65,000,000 computrons

swingset has a run policy... cosmic-swingset uses a limit on computrons per block to make such a policy.

screenshot from [swingset params](https://cosgov.org/agoric?msgType=parameterChangeProposal&network=main):
![image](https://github.com/user-attachments/assets/7c76c10c-e7b3-45d1-8835-f56bdd0073cf)

**Comment by dckc** [comment]
Note @tgrecojs reports some [interesting contract performance testing results](https://github.com/hindley-milner-systems/dapp-ertp-airdrop/issues/111#issuecomment-2479820290) for the Tribbles Airdrop contract on a local chain.

**Comment by dckc** [comment]
As discussed in [today's office hours](https://github.com/Agoric/agoric-sdk/discussions/10819)...

https://github.com/Agoric/agoric-sdk/blob/a4a4683ea0ae72ffc4a9179657437386149860bd/a3p-integration/proposals/z:acceptance/stakeBld.test.js#L37

```js
  await agdWalletUtils.broadcastBridgeAction(GOV1ADDR, {
    method: 'executeOffer',
    offer: {
      id: 'request-stake',
      invitationSpec: {
        source: 'agoricContract',
        instancePath: ['stakeBld'],
        callPipe: [['makeStakeBldInvitation']],
      },
      proposal: {
        give: {
          In: { brand: BLDBrand, value: 10n },
        },
      },
    },
  });
```

then something like...

```js
  const istBalanceAfter = await retryUntilCondition(
    async () => getBalances([GOV1ADDR], 'uist'),
    // We only check gov1's IST balance is changed.
    // Because the reclaimed amount (0,015 IST) is less than execution fee (0,2 IST)
    // we might end up with less IST than the one before reclaiming the stuck payment.
    istBalance => istBalance !== istBalanceBefore,
    'tryExitOffer did not end up changing the IST balance',
    { log: t.log, setTimeout, retryIntervalMs: 5000, maxRetries: 15 },
  );
```

see also

 - https://github.com/Agoric/agoric-sdk/issues/8963

**Comment by dckc** [comment]
In qualifying Fast USDC for release, we (@turadg , @mhofman , and co) came up with some guidelines to recommend as well as some techniques for addressing them.
_I'm saving this in **draft** form; stay tuned for refinements._

We distinguished between "resource consumption" testing and "divergence" testing. The latter (e.g. #10940 for Fast USDC) needs  _coverage_ whereas the former (#10890) needs realistic production loads.

### Resource consumption testing

While this often goes by “perf” testing, MN2 qualification is less about user-perceived latency and throughput and more about impact on long-term health of the chain, including leaked resources as well as consuming too much compute.

Structure

- Set up an “iteration” which will be repeated; all product functions of the contract SHOULD be covered
- Capture all the metrics for each iteration

Per iteration

- Exported objects
  - _no growth, or growth corresponding to product features_
  - [int] `clistEntries`
  - [int] `kernelObjects`
  - _In boot tests, available on `controller.getStats()`_
- Pending work
  - _No pending work after operation is complete_
  - [int] `promiseQueueLength`
  - [int] `kernelPromises`
  - _In boot tests, available on `controller.getStats()`_
- Vstorage
  - [int] Count of entries
  - [int] Total size
    - len(serialize(mock map))
  - _Treating append-mode as set-mode, which is what RunUtils vstorage mock does_
- Computrons
  - _Compute is consistent per iteration_
  - [int] Computrons
- Heap
  - [int] `uncompressedSize` of ‘current’ (i.e. latest) snapshot of each vat
    - Available from snapStore.listAllSnapshots()

While `local` node.js workers can be used for the first few categories for better DX (speed, debugger support...), Computrons and Heap can only be measured on XS workers.

### Divergence testing

Just needs _coverage_ to exercise all the code paths and confirm no divergence happens (with >1 validators so a3p and current multichain-testing are insufficient).

The report is to describe the coverage and show that no divergence occurred.

**Reply by mhofman** [reply]
The divergence testing in this case was required because Fast USDC is the first deployed contract exercising new cosmos features of the chain. Our JS environment is fully deterministic. golang however is specified to be non deterministic in a bunch of cases (and explicitly implements variations for non deterministic behavior). I do not expect any other contract to have to go through any requirement to test for divergence/non-determinism.

**Comment by dckc** [comment]
state of the art is probably:

 - https://github.com/Agoric/agoric-sdk/pull/11248
